vimiv/plugins/metadata_piexif.py

```diff
@@ -0,0 +1,149 @@
+# vim: ft=python fileencoding=utf-8 sw=4 et sts=4
+
+# This file is part of vimiv.
+# Copyright 2017-2023 Christian Karl (karlch) <karlch at protonmail dot com>
+# License: GNU GPL v3, see the "LICENSE" and "AUTHORS" files for details.
+
+"""Metadata plugin based on piexif (https://pypi.org/project/piexif/) backend.
+
+Properties:
+- Simple and easy to install.
+- Limited image type support (JPG and TIFF only).
+- No formatting of metadata.
+- Can only handle Exif.
+"""
+
+import contextlib
+from typing import Any, Sequence, Iterable
+
+from vimiv.imutils import metadata
+from vimiv.utils import log, lazy
+
+piexif = lazy.import_module("piexif")
+
+_logger = log.module_logger(__name__)
+
+
+class MetadataPiexif(metadata.MetadataPlugin):
+    """Provided metadata support based on piexif.
+
+    Implements `get_metadata`, `get_keys`, `copy_metadata`, and `get_date_time`.
+    """
+
+    def __init__(self, path: str) -> None:
+        self._path = path
+
+        try:
+            self._metadata = piexif.load(path)
+        except FileNotFoundError:
+            _logger.debug("File %s not found", path)
+            self._metadata = None
+        except piexif.InvalidImageDataError:
+            log.warning(
+                "Piexif only supports the file types JPEG and TIFF.<br>\n"
+                "Please use another metadata plugin for better file type support.<br>\n"
+                "For more information see<br>\n"
+                "https://karlch.github.io/vimiv-qt/documentation/exif.html",
+                once=True,
+            )
+            self._metadata = None
+
+    @property
+    def name(self) -> str:
+        """Get the name of the used backend."""
+        return "piexif"
+
+    @property
+    def version(self) -> str:
+        """Get the version of the used backend."""
+        return piexif.VERSION
+
+    def get_metadata(self, desired_keys: Sequence[str]) -> metadata.MetadataDictT:
+        """Get value of all desired keys for the current image."""
+        out = {}
+
+        # TODO: remove
+        desired_keys = [key.rpartition(".")[2] for key in desired_keys]
```

Comment:
Remove as this is filtered later on?

Reply:
Sorry for not providing a proper ToDo message. This conversation, and the subsequent are directly related to each other. Though, they have nothing to do with any sort of filtering.

Let me elaborate a bit...
pyexiv2 and piexif use different keys to index the same data. The former one used a `Group.Subgroup.Keyname` key, while the later one uses simply `Keyname` (actually, it also uses some kind of prefix, but it is off the standard).
When adding pyexiv2, we thought about backwards compatibility in the following sense:
- Backwards Compatibility Nr 1: Someone having the old piexif keys defined in their config, but wanting to use the (at that time) new pyexiv2 backend.
    - Solution: We *predict* the `Group.Subgroup` prefix to be one of the very common ones.
- Backwards Compatibility Nr 2: Someone having the new pyexiv2 keys in their config (which are the default), but wanting to use the old piexif.
    - Solution: Strip `Group.Subgroup` prefix

These two ToDos are referring to this backwards compatibility. The issue with maintaining the backwards compatibility is, in my view, that it may be a bit confusing; The keys defined in the config do not match with the ones returned by `:metadata-key-list`. Nevertheless, the widget is able to retrieve and display some data for this key. Also, the behaviour is (so far) not documented.

Personally, I do not think that is makes sense to maintain backwards compatibility Nr 1 any longer.However, for Nr 2, it probably makes sense to keep it as is. This way one does not need to modify the metadata keys in the config in order to have metadata support using the piexif based plugin.

What do you reckon?

Reply:
Ah, thanks for your clarification, I somehow thought you wanted to remove the `desired_keys` part in general.

This makes perfect sense, I agree with removing Nr. 1 and keeping Nr. 2 :+1:

Reply:
Done in 2835b92aa3a7394dd27328eea5f4309012ac51ed